x/swaprouter/client/cli/tx.go

```diff
@@ -0,0 +1,360 @@
+			if err != nil {
+				return err
+			}
+
+			txf := tx.NewFactoryCLI(clientCtx, cmd.Flags()).WithTxConfig(clientCtx.TxConfig).WithAccountRetriever(clientCtx.AccountRetriever)
+
+			txf, msg, err := NewBuildSwapExactAmountOutMsg(clientCtx, args[0], args[1], txf, cmd.Flags())
+			if err != nil {
+				return err
+			}
+
+			return tx.GenerateOrBroadcastTxWithFactory(clientCtx, txf, msg)
+		},
+	}
+
+	cmd.Flags().AddFlagSet(FlagSetSwapAmountOutRoutes())
+	flags.AddTxFlagsToCmd(cmd)
+	_ = cmd.MarkFlagRequired(FlagSwapRoutePoolIds)
+	_ = cmd.MarkFlagRequired(FlagSwapRouteDenoms)
+
+	return cmd
+}
+
+func NewBuildSwapExactAmountInMsg(clientCtx client.Context, tokenInStr, tokenOutMinAmtStr string, txf tx.Factory, fs *flag.FlagSet) (tx.Factory, sdk.Msg, error) {
+	routes, err := swapAmountInRoutes(fs)
+	if err != nil {
+		return txf, nil, err
+	}
+
+	tokenIn, err := sdk.ParseCoinNormalized(tokenInStr)
+	if err != nil {
+		return txf, nil, err
+	}
+
+	tokenOutMinAmt, ok := sdk.NewIntFromString(tokenOutMinAmtStr)
+	if !ok {
+		return txf, nil, fmt.Errorf("invalid token out min amount, %s", tokenOutMinAmtStr)
+	}
+	msg := &types.MsgSwapExactAmountIn{
+		Sender:            clientCtx.GetFromAddress().String(),
+		Routes:            routes,
+		TokenIn:           tokenIn,
+		TokenOutMinAmount: tokenOutMinAmt,
+	}
+
+	return txf, msg, nil
+}
+
+func NewBuildSwapExactAmountOutMsg(clientCtx client.Context, tokenOutStr, tokenInMaxAmountStr string, txf tx.Factory, fs *flag.FlagSet) (tx.Factory, sdk.Msg, error) {
+	routes, err := swapAmountOutRoutes(fs)
+	if err != nil {
+		return txf, nil, err
+	}
+
+	tokenOut, err := sdk.ParseCoinNormalized(tokenOutStr)
+	if err != nil {
+		return txf, nil, err
+	}
+
+	tokenInMaxAmount, ok := sdk.NewIntFromString(tokenInMaxAmountStr)
+	if !ok {
+		return txf, nil, errors.New("invalid token in max amount")
+	}
+	msg := &types.MsgSwapExactAmountOut{
+		Sender:           clientCtx.GetFromAddress().String(),
+		Routes:           routes,
+		TokenInMaxAmount: tokenInMaxAmount,
+		TokenOut:         tokenOut,
+	}
+
+	return txf, msg, nil
+}
+
+func NewCreatePoolCmd() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "create-pool [flags]",
+		Short: "create a new pool and provide the liquidity to it",
+		Long:  `Must provide path to a pool JSON file (--pool-file) describing the pool to be created`,
+		Example: `Sample pool JSON file contents:
+{
+	"weights": "4uatom,4osmo,2uakt",
+	"initial-deposit": "100uatom,5osmo,20uakt",
+	"swap-fee": "0.01",
+	"exit-fee": "0.01",
+	"future-governor": "168h"
+}
+`,
+		Args: cobra.ExactArgs(0),
+		RunE: func(cmd *cobra.Command, args []string) error {
+			clientCtx, err := client.GetClientTxContext(cmd)
+			if err != nil {
+				return err
+			}
+
+			txf := tx.NewFactoryCLI(clientCtx, cmd.Flags()).WithTxConfig(clientCtx.TxConfig).WithAccountRetriever(clientCtx.AccountRetriever)
+
+			txf, msg, err := NewBuildCreateBalancerPoolMsg(clientCtx, txf, cmd.Flags())
```

Comment:
How does this work, I thought we use the same CreatePool cli cmd for both balancer and stableswap?

Reply:
We use the same `CreatePool` keeper method. This method takes a `CreatePoolMsg` interface. Each pool has a message implementing this interface.

This is the client side of things where the message is being built. What you're referring to is chain side.